proposals/nnnn-custom-actor-executors.md

```diff
@@ -0,0 +1,593 @@
+
+  For another example, if an asynchronous function makes many calls to the same actor without any intervening suspensions, running the function explicitly on that actor's executor may eventually allow Swift to avoid a lot of switching overhead (or may even be necessary to perform those calls "atomically").
+
+This is the first proposal discussing custom executors and customization points in the Swift Concurrency runtime, and while it introduces only the most basic customization points, we are certain that it already provides significant value to users seeking tighter control over their actor's execution semantics.
+
+## Proposed solution
+
+We propose to allow actors to declare the executor they are required to execute on by declaring a nonisolated `unownedExecutor` property with the executor of their choice.
+
+## Detailed design
+
+### A low-level design
+
+The API design of executors is intended to support high-performance implementations, with an expectation that custom executors will be primarily implemented by experts. Therefore, the following design heavily prioritizes the reliable elimination of abstraction costs over most other conceivable goals. In particular, the primitive operations specified by protocols are generally expressed in terms of opaque, unsafe types which implementations are required to use correctly. These operations are then used to implement more convenient APIs as well as the high-level language operations of Swift Concurrency.
+
+### Executors
+
+First, we introduce an `Executor` protocol, that serves as the parent protocol of all the specific kinds of executors we'll discuss next. It is the simplest kind of executor that does not provide any ordering guarantees about the submitted work. It could decide to run the submitted jobs in parallel, or sequentially.
+
+This protocol has existed in Swift ever since the introduction of Swift Concurrency, however, in this proposal we revise its API to make use of the newly introduced move-only capabilities in the language. The existing `UnownedJob` API will be deprecated in favor of one accepting a move-only  `Job`. The `UnownedJob` type remains available (and equally unsafe), because today still some usage patterns are not supported by the initial revision of move-only types.
+
+The concurrency runtime uses the `enqueue(_:)` method of an executor to schedule some work onto given executor.
+
+```swift
+/// A service that can execute jobs.
+@available(SwiftStdlib 5.1, *)
+public protocol Executor: AnyObject, Sendable {
+
+  // This requirement is repeated here as a non-override so that we
+  // get a redundant witness-table entry for it.  This allows us to
+  // avoid drilling down to the base conformance just for the basic
+  // work-scheduling operation.
+  @available(SwiftStdlib 5.9, *)
+  func enqueue(_ job: __owned Job)
+
+  @available(SwiftStdlib 5.1, *)
+  @available(*, deprecated, message: "Implement the enqueue(Job) method instead")
```

Comment:
Just as a heads up, this combination of availability attributes won't work correctly in that it is silently erasing the `introduced:` version for this API on platforms where an introduced version is needed. An `@available(*, ...)` attribute always takes precedence over everything else, so unfortunately you'll need to spell out the deprecation for each platform with explicit OS versions.

I have a couple bugs I'm tracking for this. One to warn you when you combine attributes like this since it will almost certainly behave unlike what you were expected. The other is to support availability macros (`SwiftStdlib 5.1` in this case) in long form `@available` attributes.

Reply:
Actually, since `Executor` itself is introduced in `SwiftStdlib 5.1`, thankfully you don't need the `introduced` version on `enqueue` since it is inherited from the enclosing type and you can just omit `@available(SwiftStdlib 5.1, *)` instead of having the conflicting attributes.

Reply:
Ah great, thanks for spotting this!